ios/SingularBridge.m

```diff
@@ -0,0 +1,181 @@
+#elif __has_include(“RCTBridge.h”)
+#else
+#import “RCTBridge.h”
+#endif
+
+#if __has_include(<React/RCTEventDispatcher.h>)
+#import <React/RCTEventDispatcher.h>
+#elif __has_include(“RCTEventDispatcher.h”)
+#else
+#import “RCTEventDispatcher.h”
+#endif
+
+@implementation SingularBridge
+@synthesize bridge = _bridge;
+
+static NSString* apikey;
+static NSString* secret;
+static NSDictionary* launchOptions;
+static BOOL isSingularLinksEnabled = NO;
+static RCTEventEmitter* eventEmitter;
+
+// Saving the launchOptions for later when the SDK is initialize to handle Singular Links.
+// The client will need to call this method is his AppDelegate in didFinishLaunchingWithOptions.
++(void)startSessionWithLaunchOptions:(NSDictionary*)options{
+    launchOptions = options;
+}
+
+// Handling Singular Link when the app is opened from a Singular Link while it was in the background.
+// The client will need to call this method is his AppDelegate in continueUserActivity.
++(void)startSessionWithUserActivity:(NSUserActivity*)userActivity{
+    if(!isSingularLinksEnabled){
+        return;
+    }
+
+    [Singular startSession:apikey
+                   withKey:secret
+           andUserActivity:userActivity
+   withSingularLinkHandler:^(SingularLinkParams * params){
+        [SingularBridge handleSingularLinks:params];
+    }];
+}
+
+RCT_EXPORT_MODULE();
+
+- (NSArray<NSString *> *)supportedEvents
+{
+    return @[@"SingularLinkHandler"];
+}
+
+RCT_EXPORT_METHOD(init:(NSString*)apikey secret:(NSString*)secret customUserId:(NSString*)customUserId){
+    if(customUserId){
+        [Singular setCustomUserId:customUserId];
+    }
+
+    [Singular startSession:apikey withKey:secret];
+}
+
+RCT_EXPORT_METHOD(initWithSingularLinks:(NSString*)apikey
+                  secret:(NSString*)secret
+                  customUserId:(NSString*)customUserId){
+    if(customUserId){
+        [Singular setCustomUserId:customUserId];
+    }
+
+    isSingularLinksEnabled = YES;
+    eventEmitter = self;
+
+    [Singular startSession:apikey
+                   withKey:secret
+          andLaunchOptions:launchOptions
+   withSingularLinkHandler:^(SingularLinkParams * params){
+        [SingularBridge handleSingularLinks:params];
+    }];
+}
+
+RCT_EXPORT_METHOD(setCustomUserId:(NSString*)customUserId){
+    [Singular setCustomUserId:customUserId];
+}
+
+RCT_EXPORT_METHOD(unsetCustomUserId){
+    [Singular unsetCustomUserId];
+}
+
+RCT_EXPORT_METHOD(event:(NSString*)eventName){
+    [Singular event:eventName];
+}
+
+RCT_EXPORT_METHOD(eventWithArgs:(NSString*)eventName args:(NSString*)args){
+    if(!args){
+        [Singular event:eventName];
+    }
+
+    [Singular event:eventName withArgs:[SingularBridge jsonToDictionary:args]];
+}
+
+RCT_EXPORT_METHOD(revenue:(NSString*)currency amount:(double)amount){
+    [Singular revenue:currency amount:amount];
+}
+
+RCT_EXPORT_METHOD(revenueWithArgs:(NSString*)currency amount:(double)amount args:(NSString*)args){
+    if(!args){
```

Comment:
made me think... wouldn't it be better if `jsonToDictionary` would handle a nil argument? and then we can skip those `if`'s in the wrapper functions? just pass the `nil` to the native function.

Reason I think this is better is that it will be easier to debug problems this way, because there is minimum logic before calling the native method

Reply:
I agree it's more elegant